File: src/connect4/experiment.py
"""
Experiment runner for Connect 4.

Includes:
- play_one_game: Play a single game between two bots from a given state.
- ExperimentSuite: Run a student evaluator against all test evaluators
  across all board states.
"""

from typing import Optional, Any, List, Tuple, Dict
from .board import GameState
from .bots import MinimaxBot
from .rules import get_legal_moves, apply_move


def play_one_game(
    bot_a: MinimaxBot,
    bot_b: MinimaxBot,
    initial_state: Optional[GameState] = None,
) -> Optional[int]:
    """
    Play a single game and return the winner.

    bot_a plays as player 0 (whoever current_player is in initial_state).
    bot_b plays as player 1.

    Returns:
        Winner (0 or 1) or None if draw.
    """
    state = initial_state.copy() if initial_state else GameState()

    while not state.is_terminal:
        legal_moves = get_legal_moves(state)
        if not legal_moves:
            break

        if state.current_player == 0:
            move = bot_a.choose_move(state)
        else:
            move = bot_b.choose_move(state)

        if move is None:
            break

        state = apply_move(state, move)

    return state.winner


class ExperimentSuite:
    """
    Run a student evaluator against all test evaluators on all board states.

    For each (test_evaluator, board_state) pair, plays 2 games:
    one with the student going first, one with the opponent going first.
    """

    def __init__(
        self,
        evaluator: Any,
        evaluator_name: str,
        test_evaluators: Dict[str, Any],
        board_states: List[Tuple[str, GameState]],
        depth: int = 4,
    ):
        self.evaluator = evaluator
        self.evaluator_name = evaluator_name
        self.test_evaluators = test_evaluators
        self.board_states = board_states
        self.depth = depth

    def run(self):
        """Run all matches and print results in real-time."""
        num_opponents = len(self.test_evaluators)
        num_boards = len(self.board_states)
        total_games = num_opponents * num_boards * 2

        print(f"\nTesting: {self.evaluator_name}")
        print(f"{num_opponents} opponents x {num_boards} board states x 2 sides = {total_games} games")
        print("=" * 60)

        student_bot = MinimaxBot(evaluator=self.evaluator, depth=self.depth)

        # Track per-opponent results
        opponent_results = {}
        total_wins = 0
        total_losses = 0
        total_draws = 0
        game_number = 0

        for opp_name, opp_evaluator in self.test_evaluators.items():
            opp_bot = MinimaxBot(evaluator=opp_evaluator, depth=self.depth)
            opp_wins = 0
            opp_losses = 0
            opp_draws = 0

            print(f"\n  vs {opp_name}:")

            for board_name, board_state in self.board_states:
                results = []

                for student_goes_first in [True, False]:
                    game_number += 1
                    side = "O" if student_goes_first else "X"

                    if student_goes_first:
                        winner = play_one_game(student_bot, opp_bot, board_state)
                        student_won = (winner == 0)
                        opp_won = (winner == 1)
                    else:
                        winner = play_one_game(opp_bot, student_bot, board_state)
                        student_won = (winner == 1)
                        opp_won = (winner == 0)

                    if student_won:
                        results.append(f"Win as {side}")
                        opp_wins += 1
                    elif opp_won:
                        results.append(f"Loss as {side}")
                        opp_losses += 1
                    else:
                        results.append(f"Draw as {side}")
                        opp_draws += 1

                print(f"    {board_name}: {results[0]}, {results[1]}")

            total_wins += opp_wins
            total_losses += opp_losses
            total_draws += opp_draws
            opp_total = opp_wins + opp_losses + opp_draws
            opp_pct = 100 * opp_wins / opp_total if opp_total else 0
            opponent_results[opp_name] = (opp_wins, opp_losses, opp_draws)
            print(f"    Subtotal: W:{opp_wins} L:{opp_losses} D:{opp_draws} ({opp_pct:.0f}% wins)")

        # Final summary
        print("\n" + "=" * 60)
        print(f"EXPERIMENT RESULTS: {self.evaluator_name}")
        print("=" * 60)

        for opp_name, (w, l, d) in opponent_results.items():
            opp_total = w + l + d
            pct = 100 * w / opp_total if opp_total else 0
            print(f"  vs {opp_name:20s}  W:{w:2d}  L:{l:2d}  D:{d:2d}  ({pct:.0f}%)")

        print("-" * 60)
        grand_total = total_wins + total_losses + total_draws
        win_pct = 100 * total_wins / grand_total if grand_total else 0
        print(f"  {'TOTAL':23s}  W:{total_wins:2d}  L:{total_losses:2d}  D:{total_draws:2d}  ({win_pct:.0f}%)")
        print("=" * 60)

        print(f"\nGRADE GUIDE:")
        print(f"  70%+ win rate: Excellent heuristic design")
        print(f"  50%+ win rate: Good heuristic design")
        print(f"  30%+ win rate: Needs improvement")
        print(f"  <30% win rate: Significant issues")

File: engine/experiment.c
#include <stdio.h>
#include "experiment.h"
#include "rules.h"

int play_one_game(const Bot *bot_a, const Bot *bot_b, const GameState *initial) {
    GameState state;
    if (initial)
        gamestate_copy(initial, &state);
    else
        gamestate_init(&state);

    while (!state.is_terminal) {
        Move legal_moves[MAX_LEGAL_MOVES];
        int num_moves = get_legal_moves(&state, legal_moves);
        if (num_moves == 0) break;

        Move move;
        int found;
        if (state.current_player == CELL_P0)
            found = bot_choose_move(bot_a, &state, &move);
        else
            found = bot_choose_move(bot_b, &state, &move);

        if (!found) break;
        state = apply_move(&state, move);
    }

    return state.winner;
}

void run_experiment(const Bot *student_bot,
                    const Bot *opponents[], int num_opponents,
                    const GameState *board_states,
                    const char board_names[][MAX_STATE_NAME],
                    int num_boards) {
    int total_wins = 0, total_losses = 0, total_draws = 0;
    int opp_wins[MAX_OPPONENTS], opp_losses[MAX_OPPONENTS], opp_draws[MAX_OPPONENTS];
    int oi, bi;
    int total_games = num_opponents * num_boards * 2;

    printf("\nTesting: %s\n", student_bot->name);
    printf("%d opponents x %d board states x 2 sides = %d games\n",
           num_opponents, num_boards, total_games);
    printf("============================================================\n");

    for (oi = 0; oi < num_opponents; oi++) {
        opp_wins[oi] = 0;
        opp_losses[oi] = 0;
        opp_draws[oi] = 0;

        printf("\n  vs %s:\n", opponents[oi]->name);

        for (bi = 0; bi < num_boards; bi++) {
            const GameState *bs = &board_states[bi];
            int side;
            const char *results[2] = { NULL, NULL };

            for (side = 0; side < 2; side++) {
                int winner;
                int student_won, opp_won;
                const char *side_str = (side == 0) ? "O" : "X";

                if (side == 0) {
                    /* Student goes first */
                    winner = play_one_game(student_bot, opponents[oi], bs);
                    student_won = (winner == CELL_P0);
                    opp_won = (winner == CELL_P1);
                } else {
                    /* Opponent goes first */
                    winner = play_one_game(opponents[oi], student_bot, bs);
                    student_won = (winner == CELL_P1);
                    opp_won = (winner == CELL_P0);
                }

                if (student_won) {
                    opp_wins[oi]++;
                    char buf[32];
                    snprintf(buf, sizeof(buf), "Win as %s", side_str);
                    results[side] = student_won ? "Win" : "Loss";
                    /* Print inline below */
                } else if (opp_won) {
                    opp_losses[oi]++;
                } else {
                    opp_draws[oi]++;
                }

                /* We'll print after both sides */
                if (student_won)
                    results[side] = side == 0 ? "Win as O" : "Win as X";
                else if (opp_won)
                    results[side] = side == 0 ? "Loss as O" : "Loss as X";
                else
                    results[side] = side == 0 ? "Draw as O" : "Draw as X";
            }

            printf("    %s: %s, %s\n", board_names[bi], results[0], results[1]);
        }

        total_wins += opp_wins[oi];
        total_losses += opp_losses[oi];
        total_draws += opp_draws[oi];

        {
            int opp_total = opp_wins[oi] + opp_losses[oi] + opp_draws[oi];
            int opp_pct = opp_total > 0 ? (100 * opp_wins[oi] / opp_total) : 0;
            printf("    Subtotal: W:%d L:%d D:%d (%d%% wins)\n",
                   opp_wins[oi], opp_losses[oi], opp_draws[oi], opp_pct);
        }
    }

    /* Final summary */
    printf("\n============================================================\n");
    printf("EXPERIMENT RESULTS: %s\n", student_bot->name);
    printf("============================================================\n");

    for (oi = 0; oi < num_opponents; oi++) {
        int opp_total = opp_wins[oi] + opp_losses[oi] + opp_draws[oi];
        int pct = opp_total > 0 ? (100 * opp_wins[oi] / opp_total) : 0;
        printf("  vs %-20s  W:%2d  L:%2d  D:%2d  (%d%%)\n",
               opponents[oi]->name, opp_wins[oi], opp_losses[oi], opp_draws[oi], pct);
    }

    printf("------------------------------------------------------------\n");
    {
        int grand_total = total_wins + total_losses + total_draws;
        int win_pct = grand_total > 0 ? (100 * total_wins / grand_total) : 0;
        printf("  %-23s  W:%2d  L:%2d  D:%2d  (%d%%)\n",
               "TOTAL", total_wins, total_losses, total_draws, win_pct);
    }
    printf("============================================================\n");
}
